Sparse: Implement simple unary ufuncs operators

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #68887

Closes #46988, closes #46987, closes #46761

By "simple" I mean operators that map 0->0 so we can implement it by
just re-dispatching on the values tensor. That does mean we have `sin`
but not `cos` for example, but without fill value support this is the
best that can be done.

Most of these don't support autograd because the derivative formulas
use unsupported operators.

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk

Differential Revision: [D32734911](https://our.internmc.facebook.com/intern/diff/D32734911)

peterbell10 committed:
Sparse: Implement simple unary ufuncs operators

By "simple" I mean operators that map 0->0 so we can implement it by
just re-dispatching on the values tensor. That does mean we have `sin`
but not `cos` for example, but without fill value support this is the
best that can be done.

Most of these don't support autograd because the derivative formulas
use unsupported operators.

[ghstack-poisoned]

peterbell10 committed:
Update on "Sparse: Implement simple unary ufuncs operators"


By "simple" I mean operators that map 0->0 so we can implement it by
just re-dispatching on the values tensor. That does mean we have `sin`
but not `cos` for example, but without fill value support this is the
best that can be done.

Most of these don't support autograd because the derivative formulas
use unsupported operators.

[ghstack-poisoned]

peterbell10 committed:
Update on "Sparse: Implement simple unary ufuncs operators"


By "simple" I mean operators that map 0->0 so we can implement it by
just re-dispatching on the values tensor. That does mean we have `sin`
but not `cos` for example, but without fill value support this is the
best that can be done.

Most of these don't support autograd because the derivative formulas
use unsupported operators.

[ghstack-poisoned]

peterbell10 committed:
Update on "Sparse: Implement simple unary ufuncs operators"


By "simple" I mean operators that map 0->0 so we can implement it by
just re-dispatching on the values tensor. That does mean we have `sin`
but not `cos` for example, but without fill value support this is the
best that can be done.

Most of these don't support autograd because the derivative formulas
use unsupported operators.

[ghstack-poisoned]

IvanYashchuk commented on aten/src/ATen/native/sparse/SparseUnaryOps.cpp:
```diff
@@ -0,0 +1,147 @@
+
+  auto input_values = input.values();
+  auto result_values = result._values();
+  result_values.resize_(input_values.sizes());
```
Resizing indices&values of sparse tensors doesn't work like this. In this case, `result_values` is resized, but `result._values()` still has the original size. You need to use `SparseTensorImpl::set_indices_and_values_unsafe` to really set the newly resized tensors as indices and values. See also https://github.com/pytorch/pytorch/issues/63549

```python
In [1]: import torch

In [2]: a = torch.zeros(3, 3).to_sparse()

In [3]: a_indices = a._indices()

In [4]: a.shape
Out[4]: torch.Size([3, 3])

In [5]: a_indices.shape
Out[5]: torch.Size([2, 0])

In [6]: a_indices.resize_((2, 3))
Out[6]:
tensor([[140315210074560, 140315191773632, 140315210132992],
        [140315210045888, 140315210131968, 140315210430464]])

In [7]: a_indices.shape
Out[7]: torch.Size([2, 3])

In [8]: a._indices().shape
Out[8]: torch.Size([2, 0])
```

peterbell10 replied:
I've changed it to access `values` and `indices` without dispatching, will that work?

IvanYashchuk replied:
I guess it should work now. Best to have a test passing "out=" tensor with the wrong size to check.

peterbell10 committed:
Update on "Sparse: Implement simple unary ufuncs operators"


Closes #46988, closes #46987, closes #46761

By "simple" I mean operators that map 0->0 so we can implement it by
just re-dispatching on the values tensor. That does mean we have `sin`
but not `cos` for example, but without fill value support this is the
best that can be done.

Most of these don't support autograd because the derivative formulas
use unsupported operators.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

peterbell10 commented on test/test_sparse.py:
```diff
@@ -1966,17 +1966,14 @@ def is_integral(dtype):
         self.assertEqual(expected_output, sparse_tensor.log1p().to_dense())
         if is_integral_dtype:
-            with self.assertRaisesRegex(RuntimeError, "log1p: result type cannot be Integral, got:"):
+            with self.assertRaisesRegex(RuntimeError, "result type .* can't be cast to"):
```
The OpInfo tests are extremely minimal, unfortunately.

peterbell10 committed:
Update on "Sparse: Implement simple unary ufuncs operators"


Closes #46988, closes #46987, closes #46761

By "simple" I mean operators that map 0->0 so we can implement it by
just re-dispatching on the values tensor. That does mean we have `sin`
but not `cos` for example, but without fill value support this is the
best that can be done.

Most of these don't support autograd because the derivative formulas
use unsupported operators.

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

cpuhrsch reviewed (approved):
LGTM

I assume CSR and autograd is next?

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32706197).

cpuhrsch commented:
@peterbell10 - could you rebase please?

peterbell10 committed:
Update on "Sparse: Implement simple unary ufuncs operators"


Closes #46988, closes #46987, closes #46761

By "simple" I mean operators that map 0->0 so we can implement it by
just re-dispatching on the values tensor. That does mean we have `sin`
but not `cos` for example, but without fill value support this is the
best that can be done.

Most of these don't support autograd because the derivative formulas
use unsupported operators.

cc nikitaved pearu cpuhrsch IvanYashchuk

Differential Revision: [D32706197](https://our.internmc.facebook.com/intern/diff/D32706197)

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32706197).

IvanYashchuk commented:
[macos-10-15-py3-lite-interpreter-x86-64 / build CI job failed](https://github.com/pytorch/pytorch/runs/4364687526) with missing symbols errors.

<details>
  <summary>Logs</summary>

```
2021-11-30T09:24:45.4498310Z Undefined symbols for architecture x86_64:
2021-11-30T09:24:45.4498950Z   "at::native::abs_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4499870Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__abs(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4500760Z   "at::native::erf_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4501680Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__erf(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4502590Z   "at::native::sgn_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4503570Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__sgn(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4504490Z   "at::native::sin_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4505390Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__sin(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4506300Z   "at::native::tan_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4507200Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__tan(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4508080Z   "at::native::abs_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4508950Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__abs_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4509840Z   "at::native::asin_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4510760Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__asin(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4511670Z   "at::native::atan_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4512580Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__atan(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4513490Z   "at::native::ceil_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4514730Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__ceil(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4515630Z   "at::native::erf_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4516510Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__erf_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4517370Z   "at::native::sgn_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4518220Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__sgn_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4519110Z   "at::native::sign_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4520170Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__sign(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4521100Z   "at::native::sin_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4521970Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__sin_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4522860Z   "at::native::sinh_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4523770Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__sinh(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4524670Z   "at::native::sqrt_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4525570Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__sqrt(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4526450Z   "at::native::tan_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4527320Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__tan_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4528210Z   "at::native::tanh_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4529110Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__tanh(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4530010Z   "at::native::asin_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4530880Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__asin_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4531770Z   "at::native::asinh_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4532690Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__asinh(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4533670Z   "at::native::atan_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4534560Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__atan_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4535460Z   "at::native::atanh_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4536380Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__atanh(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4537270Z   "at::native::ceil_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4538150Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__ceil_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4539050Z   "at::native::expm1_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4539970Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__expm1(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4540890Z   "at::native::floor_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4541800Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__floor(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4542710Z   "at::native::isinf_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4543620Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__isinf(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4544530Z   "at::native::isnan_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4545440Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__isnan(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4546630Z   "at::native::log1p_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4547550Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__log1p(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4548460Z   "at::native::round_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4549370Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__round(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4550260Z   "at::native::sign_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4551270Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__sign_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4552180Z   "at::native::sinh_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4553060Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__sinh_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4553940Z   "at::native::sqrt_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4554810Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__sqrt_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4555680Z   "at::native::tanh_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4556550Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__tanh_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4557440Z   "at::native::trunc_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4558350Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__trunc(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4559260Z   "at::native::asinh_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4560140Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__asinh_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4561020Z   "at::native::atanh_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4561910Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__atanh_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4562810Z   "at::native::erfinv_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4563810Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__erfinv(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4564740Z   "at::native::expm1_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4565630Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__expm1_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4566520Z   "at::native::floor_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4567400Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__floor_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4568280Z   "at::native::log1p_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4569160Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__log1p_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4570050Z   "at::native::round_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4570950Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__round_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4571850Z   "at::native::trunc_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4572740Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__trunc_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4573670Z   "at::native::abs_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4574680Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_abs_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4575690Z   "at::native::erf_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4576700Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_erf_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4578010Z   "at::native::erfinv_sparse_(at::Tensor&)", referenced from:
2021-11-30T09:24:45.4578900Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__erfinv_(at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4579850Z   "at::native::sgn_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4580850Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_sgn_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4581830Z   "at::native::signbit_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4582930Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__signbit(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4584060Z   "at::native::sin_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4585170Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_sin_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4586350Z   "at::native::tan_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4587450Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_tan_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4588850Z   "at::native::asin_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4589920Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_asin_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4591000Z   "at::native::atan_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4592100Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_atan_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4593200Z   "at::native::ceil_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4594790Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_ceil_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4595880Z   "at::native::isneginf_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4596910Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__isneginf(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4597940Z   "at::native::isposinf_sparse(at::Tensor const&)", referenced from:
2021-11-30T09:24:45.4599040Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__isposinf(at::Tensor const&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4600060Z   "at::native::sign_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4601160Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_sign_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4602250Z   "at::native::sinh_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4603350Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_sinh_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4604600Z   "at::native::sqrt_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4605640Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_sqrt_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4606730Z   "at::native::tanh_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4607820Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_tanh_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4608990Z   "at::native::asinh_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4610100Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_asinh_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4611550Z   "at::native::atanh_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4612660Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_atanh_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4613770Z   "at::native::expm1_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4614970Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_expm1_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4616060Z   "at::native::floor_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4617540Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_floor_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4619860Z   "at::native::log1p_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4622170Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_log1p_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4624270Z   "at::native::round_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4626150Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_round_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4627240Z   "at::native::trunc_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4628340Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_trunc_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4629850Z   "at::native::erfinv_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4631070Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_erfinv_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4632530Z   "at::native::nan_to_num_sparse(at::Tensor const&, c10::optional<double>, c10::optional<double>, c10::optional<double>)", referenced from:
2021-11-30T09:24:45.4634050Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__nan_to_num(at::Tensor const&, c10::optional<double>, c10::optional<double>, c10::optional<double>) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4635770Z   "at::native::nan_to_num_sparse_(at::Tensor&, c10::optional<double>, c10::optional<double>, c10::optional<double>)", referenced from:
2021-11-30T09:24:45.4637320Z       at::(anonymous namespace)::(anonymous namespace)::wrapper__nan_to_num_(at::Tensor&, c10::optional<double>, c10::optional<double>, c10::optional<double>) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4638570Z   "at::native::signbit_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4639700Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_signbit_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4640910Z   "at::native::isneginf_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4642070Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_isneginf_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4643140Z   "at::native::isposinf_sparse_out(at::Tensor const&, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4644270Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_isposinf_out_out(at::Tensor const&, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4645580Z   "at::native::nan_to_num_sparse_out(at::Tensor const&, c10::optional<double>, c10::optional<double>, c10::optional<double>, at::Tensor&)", referenced from:
2021-11-30T09:24:45.4647160Z       at::(anonymous namespace)::(anonymous namespace)::wrapper_out_nan_to_num_out_out(at::Tensor const&, c10::optional<double>, c10::optional<double>, c10::optional<double>, at::Tensor&) in RegisterSparseCPU.cpp.o
2021-11-30T09:24:45.4648350Z ld: symbol(s) not found for architecture x86_64
2021-11-30T09:24:45.4651440Z clang: error: linker command failed with exit code 1 (use -v to see invocation)
2021-11-30T09:24:45.4652560Z ninja: build stopped: subcommand failed.
```
  
</details>

cpuhrsch commented:
@pytorchbot ciflow rerun

cpuhrsch commented:
This adds a new file that wasn't caught by some of the grep based build tools.

cpuhrsch commented:
Updating [tools/build_variables.bzl](https://github.com/pytorch/pytorch/blob/master/tools/build_variables.bzl#L1230-L1231) should be enough to resolve this @peterbell10

peterbell10 committed:
Update on "Sparse: Implement simple unary ufuncs operators"


Closes #46988, closes #46987, closes #46761

By "simple" I mean operators that map 0->0 so we can implement it by
just re-dispatching on the values tensor. That does mean we have `sin`
but not `cos` for example, but without fill value support this is the
best that can be done.

Most of these don't support autograd because the derivative formulas
use unsupported operators.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

cpuhrsch commented:
@pytorchbot ciflow rerun

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32734911).
